File: tree/java/leetcode226_Invert_Binary_Tree.java
//Invert a binary tree.
//
//Example:
//
//Input:
//
//     4
//   /   \
//  2     7
// / \   / \
//1   3 6   9
//Output:
//
//     4
//   /   \
//  7     2
// / \   / \
//9   6 3   1

package wenxinjie2;

import java.util.Stack;

public class leetcode226_Invert_Binary_Tree {
    public TreeNode invertTree(TreeNode root) {
        Stack<TreeNode> stack = new Stack<TreeNode>();
        stack.push(root);
        while (!stack.isEmpty()){
            TreeNode node = stack.pop();
            if (node != null){
                TreeNode tem = node.left;
                node.left = node.right;
                node.right = tem;
                stack.push(node.left);
                stack.push(node.right);
            }
        }
        return root;
    }
}

//Time: O(n)
//Space: O(n)
//Difficulty: easy

File: tree/python/leetcode226_Invert_Binary_Tree.py
# Invert a binary tree.

# Example:

# Input:

#      4
#    /   \
#   2     7
#  / \   / \
# 1   3 6   9
# Output:

#      4
#    /   \
#   7     2
#  / \   / \
# 9   6 3   1


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def invertTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        if not root: return None
        queue = [root]    
        while queue:
            node = queue.pop()
            if node:
                node.left, node.right = node.right, node.left
                queue += [node.left,node.right]
        return root 
            
# Time: O(n)
# Space: O(n)
# Difficulty: easy